### ISSUE TITLE
Unable to sync multiple timers.

### ISSUE THREAD

**Issue by SteveRMann** [EasyG0ing1/BlockNot#11]
I am evaluating some of the delay() replacements in the Arduino Library manager.  
My test is running on a Wemos D1 Mini with two LEDs.  The object is to blink the LEDs in the pattern shown in the attached image.

![Async](https://user-images.githubusercontent.com/8091425/124788684-94493680-df17-11eb-8ea7-cead78de586b.jpg)

Both LEDs start at T0, and at T6.0 they both should start together again.
Try as I can, I can't get the two timers to begin together.
Have I missed something in the READ.me?

Here are my notes:
There is no way to create the timer objects without starting them.
Missing is a NO_RESET method for FIRST_TRIGGER.
If it is to be readable, where is the start and stop methods?

Here is my sketch:

```
/*
   BlockNot-BilnkAsync.ino
*/

#include <BlockNot.h>

BlockNot yellowLED_onTime(5000);
BlockNot yellowLED_offTime(1000);
BlockNot blueLED_onTime(500);
BlockNot blueLED_offTime(2500);

const int LED_ON = 1;
const int LED_OFF = 0;
const int LED_YELLOW_pin = D2;          //Data pins on Wemos D1 Mini
const int LED_BLUE_pin = D1;


void setup() {
  Serial.begin(115200);
  while (!Serial);
  Serial.println();
  Serial.println(F("BlockNot-BilnkAsync.ino"));
  Serial.println();

  pinMode(LED_YELLOW_pin, OUTPUT);
  pinMode(LED_BLUE_pin, OUTPUT);

  yellowLED_onTime.reset();             //Start the timers
  blueLED_onTime.reset();
}

void loop() {
  if (yellowLED_onTime.FIRST_TRIGGER) yellow_OFF();  //Yellow ON time expired, turn it off
  if (yellowLED_offTime.FIRST_TRIGGER) yellow_ON();  //Yellow OFF time expired, turn it on
  if (blueLED_onTime.FIRST_TRIGGER) blue_OFF();      //Blue ON time expired, turn it off
  if (blueLED_offTime.FIRST_TRIGGER) blue_ON();      //Blue OFF time expired, turn it on
}


void yellow_ON() {
  //Turn on the YELLOW LED then restart the ON timer.
  digitalWrite(LED_YELLOW_pin, LED_ON);     //Turn on the yellow LED
  yellowLED_onTime.reset();                 //Start the ON timer
}

void yellow_OFF() {
  //Turn off the YELLOW LED then restart the OFF timer.
  digitalWrite(LED_YELLOW_pin, LED_OFF);    //Turn off the yellow LED
  yellowLED_offTime.reset();                //Start the OFF timer
}

void blue_ON() {
  //Turn on the BLUE LED then restart the ON timer.
  digitalWrite(LED_BLUE_pin, LED_ON);     //Turn on the blue LED
  blueLED_onTime.reset();                 //Start the ON timer
}

void blue_OFF() {
  //Turn off the BLUE LED then restart the OFF timer.
  digitalWrite(LED_BLUE_pin, LED_OFF);    //Turn off the blue LED
  blueLED_offTime.reset();                //Start the OFF timer
}
```

**Comment by EasyG0ing1** [comment]
@SteveRMann - Interesting ... you are correct. The timers technically start the instant they are instantiated. 

"Starting and stopping" can be done with the .DISABLE and .ENABLE commands ... but even if you have, for example, two timers that you create both set for 3,000 milliseconds, then you disable them and reset them, you can still only enable one timer at a time because that's the nature of how the microcontroller works ... so even if you had back to back timer.ENABLE commands, there would be some delay between the first one being enabled and then the second so your synchronization problem might require something like a Raspberry Pi Pico which has two isolated CPU cores ... and you could put a routine in each thread that started their timers at the exact same millisecond ... but in a single-core microcontroller, you're going to have to figure out how much time difference there is between the starting of the first-timer and the starting of the second timer, then you would simply add that time delta to the first-timer, but then after it triggers the first time, and is in sync with timer 2, you would need to change its duration back to match Timer 2 ... if that makes any sense...

I'll play with it and post some code in a little bit if I'm successful.

**Comment by EasyG0ing1** [comment]
@SteveRMann - The reason why there isn't a NO_RESET option for the .FIRST_TRIGGERED command is because, after the FIRST_TRIGGER fires, it won't fire again until you manually reset the timer with a .RESET ... FIRST_TRIGGER is hardcoded to NOT reset ... so if we had it capable of accepting a NO_RESET option, it would be redundant.

**Comment by EasyG0ing1** [comment]
@SteveRMann - Also, have you tried the resetAllTimers() method? Because when you call that method, it resets every timer you have instantiated and it sets all of their start times to the same value - which technically puts them in sync with one 
 another ... so if you had say two timers set for a 1300 millisecond duration, then you called the resetAllTimers() method,  then those timers could technically trigger at the exact same moment in time - that is if you could actually call the TRIGGERED method at the exact same moment in time which can't really happen in a single-core microcontroller.

**Comment by EasyG0ing1** [comment]
@SteveRMann - this code is pretty close to what you're after...

```
#include <Arduino.h>
#include <BlockNot.h>

BlockNot blueONTimer(2500);
BlockNot blueOFFTimer(500);
BlockNot redTimer(1000);

const int PIN_B = 10; 
const int PIN_R = 9;

#define BLUE_ON digitalWrite(PIN_B, LOW);
#define BLUE_OFF digitalWrite(PIN_B, HIGH);
#define RED_ON digitalWrite(PIN_R, LOW);
#define RED_OFF digitalWrite(PIN_R, HIGH);

void setup() {
    Serial.begin(115200);
    Serial.println(F("\nReady!\n"));
    pinMode(PIN_B, OUTPUT);
    pinMode(PIN_R, OUTPUT);
    resetAllTimers();
}


void loop() {
    static int redTriggerCount = 0;
    if (blueONTimer.TRIGGERED) {
        BLUE_ON;
        blueOFFTimer.RESET;
    }
    if (blueOFFTimer.TRIGGERED) {
        BLUE_OFF;
    }
    if (redTimer.TRIGGERED) {
        if (redTriggerCount == 0 || redTriggerCount == 6) {
            RED_ON;
        }
        else if ( redTriggerCount == 5 ) {
            RED_OFF;
        }
        redTriggerCount++;
        if (redTriggerCount > 6) {
            redTriggerCount = 0;
        }
    }
}

```

**Comment by SteveRMann** [comment]
First, I don't need a lesson on how a microprocessor can only do one thing art a time.  When I say "at the same time" or "in sync", I mean within a few instruction cycles.
"this code is pretty close to what you're after..."  No, it isn't. After five minutes the LED pattern doesn't resemble anything near synchronous.
Using one timer to trigger the other is missing the point.  While it may reproduce my example pattern, the timers should be consistently repeatable _independent of each other_, and regardless how many timers you are running  (within hardware or memory constraints).
Of the ten libraries I am experimenting with, four of them meet this criteria and are still in sync after 18 to 24 hours of operation.

**Comment by EasyG0ing1** [comment]
@SteveRMann  - All I can tell you is that BlockNot is nothing more than a library that records the microcontrollers current millis() when the library is instantiated, then when your code makes subsequent calls to the instantiated library, all it does is take another reading of millis() and it does the math to figure out how many millis() has elapsed since the last time a call was made to the object. There are additional methods in there that give the developer some different info like how many milliseconds until the next trigger event etc. etc. 

I never claimed anything about BlockNot other than it is a friendly and simple way of doing non-blocking timers as an alternative to manually creating a variable that first records the millis() then having to perform math in your code to see how many millis() have passed since you made the initial recording. 

I've never claimed it to be a library to be used for precision timing, as I would never rely on an Arduino's millis() function as a means of doing precision timing. There are far more reliable ways to do precision timing as I'm sure you are well aware.

I'll keep this open for your last comment, or if you have nothing else to say within a day or so,  I'll close this issue.

**Comment by SteveRMann** [comment]
I am not looking for anything- I am simply comparing the libraries.  Precision was never in the testing criteria, but when two independent timers using the same millis() clock can't keep synchronicity with each other it fails the testing criteria.

Here are my notes regarding BlockNot:
BlockNot
This library is a recent addition to the Arduino Library Manager, so maybe still under development? This is by far one of the best documented libraries I have ever seen. It get an "B" for documentation. (It could have earned an "A" but for the unreadable illustration and rambling nature of the descriptions.  Light yellow on white is simply invisible).  This library has a NO_RESET method that stops the timer after it expires, making it a one-shot.  It's the only timer library with this needed feature.
My test code: BlockNot-BilnkAsync.ino -- I could not get the timers to sync. The resetAllTimers() method should start them at the same time but I don't see it.

Here is my testing protocol:

The timers should operate independently of each other and regardless how many timers are running (within hardware or memory constraints).  To pass this test the timers should be mostly independent from each other.  In other words, using one timer to trigger another is not independent because it would not work with timers with no repeating coincidence. Adding a third of forth timer should have no effect overall. Five of the ten libraries meet this criteria. (Note, I have not yet devised a test protocol for testing no repeating coincidence.)

In order to replace delay() function the replacement library should be as easy to implement delay(), or at least more straightforward than implementing millis().  All of the tested libraries fail this criteria. One spectacularly as the user needs to write a lambda in their code. Even experienced Arduino programmers are stumped by lambdas.

This is all still a work in progress and nowhere near ready to publish.  You are welcome to submit your solution for testing if I have missed something in my code. I really wanted BlockNot to be successful in the testing because of the outstanding documentation.

**Comment by EasyG0ing1** [comment]
@SteveRMann Documentation is easily fixed... though I'm not really sure who you're publishing for and who your audience is. Nonetheless, I changed the graphic and cut out a lot of text from the readme to cut back on much of the "rambling" ... which was a fair point made ... documentation should be efficient, I agree.

And as far as usability, It's impossible to create a function that works "as easily" as delay(); because delay does ONE THING - it blocks code execution for the duration of the argument passed into the method.

Honestly, I have never used nor looked at any other timer libraries. I developed BlockNot for myself and as it became a staple library for me, I decided to publish it in the hopes that someone else might find it as useful as I have.

But I'm not sure how you can get any easier than declaring the timer and using it with simple words and simple code. 

Making a NON-BLOCKING  timer library that is as "simple" to use as delay() would be impossible really ... the very nature of non-blocking REQUIRES interaction with code at more than one point in your code ... it's all about non-stop code flow ... so the ONLY WAY to have a timer that does not block, is to have a timer that gets defined, then evaluated at some point in the code. A one-shot command simply could never work with exception  MAYBE in a true SMP environment of which no Arduino that I know of has and no talk of it ever happening anytime soon either. The Pi Pico has a second core but even that is a far cry from SMP.

Interrupt Handlers would be different of course, but those technically aren't timers... though the DS3232 has nice "alarm" features that work in a similar fashion ... they have to be set then checked and I think they have one or two that can throw an interrupt then be handled by an interrupt handler ... but even interrupt handlers are by no means "simple" when your standard for simple is delay(milliseconds);

Now that I'm thinking about it ... I think I'm going to try your timer sync program using a DS3232 and the alarm timers and see how that goes.

You've given me a lot to think about in terms of how a timer should behave and what people might be expecting when they use this library. I'm definitely going to look into ways to improve the library and for the first time, I think I'll even look at other libraries to see how other developers do them.

I do appreciate your feedback ... though to be honest with you... I didn't understand where your seemingly combative attitude was coming from ... since I'm only trying to help ... you seemed insulted at some of the things I said and I'm not sure why since I never insulted you. When someone I don't know opens an issue and describes a problem ... you can't fault me for stating things that are true ... having no idea what you're knowledge is or background is not a crime on my part ... what I wrote was perfectly sensical and non-assumptive and certainly not insulting.

**Comment by SteveRMann** [comment]
Sorry. I was probably using my Quora voice. (You _did_ try to school me on how microprocessors work).

As I said, my testing protocol is a work in progress.  On reflection, I will have to revisit my whole testing criteria because even those that did pass would probably fail with multiple timers with no repeating coincidence. I unwittingly made the timers dependent on each other thus violating my own criteria.

DISCLAIMER: I am not a C++ expert and don't pretend to be one.  I have only written one library for Arduino a year ago. I do not grok lambda. (Yes, one library was based on macros using lambda- hardly what I would call easy to implement).  I got into this evaluation because a library that I did fork and had been using myself became hopelessly useless when adding a second instance of the timer into my code.  I (still) can't grok why timers based on the same clock (millis) can't stay in sync.

**Comment by EasyG0ing1** [comment]
@SteveRMann  I often forget that C++ even has lambda ... I use it in Java all the time ... and some of it even makes sense to me after all these years since they introduced it.  lol

I am - at this moment, looking at my 4 channel scope, each channel connected to a pin on an Arduino nano. Each pin has a BlockNot timer on it issuing digitalWrites HIGH and LOW alternating each time a given timer triggers ....

Each timer is at a different value starting at 50 milliseconds then 75, 100, 125 ...

And scopes don't lie ... these timers are in perfect sync and they aren't drifting at all

**Comment by EasyG0ing1** [comment]
@SteveRMann and I wasn't "schooling" you on microprocessors ... I was stating the obvious because it was relevant to what I was reasoning out in the discussion .... but I will admit ... when I read your code ... I did assume that you might be ... a little new to the Arduino world ... no offense by that ... your code is ... thorough ... though it lacks many common elements that would make it much more efficient and streamlined ... so I did assume a beginner to some degree ... but ... you can't really fault me for that, from my perspective, you are a perfect stranger creating an issue in this library and I only had your words to go on when I replied ... and nothing in your post said anything about your vast experience with microcontrollers ... so my comments were not insults ... I am certainly not the kind of person to insult anyone ... unless they've earned it of course.

**Comment by EasyG0ing1** [comment]
@SteveRMann I don't know if you have a scope or not, but this one has a web server in it that provides full access to all of the scopes settings along with a nice live view of the screen .... If you're interested, it wouldn't take me long at all to set you up with a virtual machine connected to a nano with 4 pins connected to the scope and you could remote access and test all the timers you want and see what they look like on the scope. Coding is NOT my bread and butter ... Network Engineering is where I've made my living ... so remote access is old hat for me.

**Comment by EasyG0ing1** [comment]
@SteveRMann 
And the reason why I decided to look at the timers using a scope is that something just didn't quite feel right about the test you were doing ... and I can't put my finger on it yet because conceptually it's a valid test ... but the notion that the timers could drift at all when they rely solely on the Arduino's millis() function just didn't sit right with me so I had to see for myself and I knew the scope would show exactly what's going on ... I am going to look at your code next with the scope and see if it sheds any light on why the timings are drifting ... cause I saw it too even in my own code ... I think we're both doing something wrong or assuming something that isn't right ...

**Comment by SteveRMann** [comment]
Show me the code and I'll try to duplicate the results.

I do use lambdas when I have no choice, but when the compiled code is not different from multi-line code to do the same thing, I see no point in making code harder to read.

**Comment by SteveRMann** [comment]
> I don't know if you have a scope or not,

I have a 2-channel Tektronics tube-type scope.  I've been thinking of buying something smaller because the tek takes up half of my bench space. And it isn't a storage scope.  I don't want to spend a lot because I don't need a scope often (IOT and Arduino programming is my hobby). Do you have any recommendations?

**Comment by SteveRMann** [comment]
Do you want to take this conversation out of the git?

**Comment by EasyG0ing1** [comment]
@SteveRMann In this video, I screen recorded the scope first sample is YOUR code copied straight off this page. I changed the pinouts so that they hit the scope and I had to change the time values because scopes suck when taking samples over long periods of time ... so I just took two zero's off the end of each timer.

Second sample in video is MY CODE copied and pasted from above with same modifications made...

Third sample is just the Blink Party example sketch in the BlockNot examples only I changed the time values.

[Here is the video](https://youtu.be/N7mIdYCMY_k)

**Comment by EasyG0ing1** [comment]
@SteveRMann  That's another thing ... when you look at my Blink Party example sketch and drop some LEDs on pins 9-12 on an Uno or a nano ... and run the sketch ... the LEDS do not change timing at all ... they stay consistent.

**Comment by EasyG0ing1** [comment]
@SteveRMann Yeah we probably should take this conversation out of git ... you can email me at sims.mike@gmail.com and we can either keep it in email or go to Messenger or whatever you're comfortable with.

**Comment by EasyG0ing1** [comment]
@SteveRMann 

Check out this code ... timers are now in sync...

```
#include <Arduino.h>
#include <BlockNot.h>

BlockNot blueTimer(500);
BlockNot redTimer(5000);

const int PIN_B = 12;
const int PIN_R = 11;

bool redON = false;
bool blueON = false;

void setRed(bool state) {
    digitalWrite(PIN_R, (state? LOW : HIGH));
}

void setBlue(bool state) {
    digitalWrite(PIN_B, (state? LOW : HIGH));
}

void setup() {
    Serial.begin(115200);
    Serial.println(F("\nReady!\n"));
    pinMode(PIN_B, OUTPUT);
    pinMode(PIN_R, OUTPUT);
    resetAllTimers();
}

void loop() {
    if (blueTimer.TRIGGERED) {
        blueON = !blueON;
        setBlue(blueON);
        if (blueON) {
            blueTimer.setDuration(500);
        }
        else {
            blueTimer.setDuration(2500);
        }
    }

    if (redTimer.TRIGGERED) {
        redON = !redON;
        setRed(redON);
        if (redON) {
            redTimer.setDuration(5000);
        }
        else {
            redTimer.setDuration(1000);
        }
    }
}

```

**Comment by EasyG0ing1** [comment]
@SteveRMann As I thought about it ... having two timers, one for on and one for off simply cant work unless you can start the off timer at the instant that the LED turns off.and then you would have to use the trigger event of the off timer to reset the on timer ...

So I guess the way to do it with TWO timers per LED, would be to turn the LED ON in the OFF timers trigger event, then RESET the on timer ... then when the on timer triggers, turn the LED off then RESET the off timer ... BUT ... and here is where it gets messy ... with one-timer being shorter duration than the other .... when the longer timer has not yet triggered, the shorter timer WILL TRIGGER and screw everything up. so then that means that you would have to DISABLE the other timer while the one is running but then you won't get a trigger event from it because it's disabled ...

The bottom line is ... having an on-timer and an off-timer simply won't work ... these are not serial chained timers like with a sprinkler system ... these things are always running and so you have to code for that ....

**Comment by EasyG0ing1** [comment]
@SteveRMann  OK, I was wrong about the shorter timer screwing things up ... all we gotta do is make sure that the state of the LED is know and don't let that other timer run code while the led is not in the state necessary for that code ...

This will explain and you'll appreciate it because in terms of readability ... I think it will exceed your expectations lol 

```
#include <Arduino.h>
#include <BlockNot.h>

BlockNot blueTimerON(500);
BlockNot blueTimerOFF(2500);
BlockNot redTimerON(5000);
BlockNot redTimerOFF(1000);

const int BLUE_PIN = 12;
const int RED_PIN = 11;

bool redON = false;
bool blueON = false;

bool blueIsOff() { return !blueON; }

bool blueIsOn() { return blueON; }

bool redIsOff() { return !redON; }

bool redIsOn() { return redON; }

void turnRedOn() {
    digitalWrite(RED_PIN, HIGH);
    redTimerON.RESET;
    redON = true;
}

void turnRedOff() {
    digitalWrite(RED_PIN, LOW);
    redTimerOFF.RESET;
    redON = false;
}

void turnBlueOn() {
    digitalWrite(BLUE_PIN, HIGH);
    blueTimerON.RESET;
    blueON = true;
}

void turnBlueOff() {
    digitalWrite(BLUE_PIN, LOW);
    blueTimerOFF.RESET;
    blueON = false;
}

void setup() {
    Serial.begin(115200);
    Serial.println(F("\nReady!\n"));
    pinMode(BLUE_PIN, OUTPUT);
    pinMode(RED_PIN, OUTPUT);
    resetAllTimers();
    turnBlueOn();
    turnRedOn();
}

void loop() {

    if (blueTimerON.TRIGGERED && blueIsOn()) {
        turnBlueOff();
    }

    if (blueTimerOFF.TRIGGERED && blueIsOff()) {
        turnBlueOn();
    }

    if (redTimerON.TRIGGERED && redIsOn()) {
        turnRedOff();
    }

    if (redTimerOFF.TRIGGERED && redIsOff()) {
        turnRedOn();
    }
}

```

and THAT is all she wrote!  It works perfectly!

**Comment by EasyG0ing1** [comment]
And [here's a short video](https://youtu.be/A6iquZSWSak) demonstrating that it works using both an on and off timer for each LED.

**Comment by EasyG0ing1** [comment]
@SteveRMann  If you're ok with it, I want to add this sketch to the examples in the library. I'm going to add a start and stop method as well and push out an update.

**Comment by SteveRMann** [comment]
Sure- it's your library.

**Comment by EasyG0ing1** [comment]
Issue resolved
